test/unit/org/apache/cassandra/index/sai/disk/io/TrackingIndexFileUtils.java

```diff
@@ -19,78 +19,114 @@
 package org.apache.cassandra.index.sai.disk.io;
 
 import java.io.IOException;
-import java.lang.reflect.Field;
-import java.lang.reflect.Modifier;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
 
 import com.google.common.base.Throwables;
 
+import org.junit.Assert;
+
 import org.apache.cassandra.index.sai.utils.IndexFileUtils;
 import org.apache.cassandra.io.util.FileHandle;
 import org.apache.cassandra.io.util.SequentialWriterOption;
-
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.fail;
+import org.apache.lucene.index.CorruptIndexException;
 
 public class TrackingIndexFileUtils extends IndexFileUtils
 {
     private final Map<TrackingIndexInput, String> openInputs = Collections.synchronizedMap(new HashMap<>());
+    private final Set<TrackingIndexInput> closedInputs = new HashSet<>();
```

Comment:
bug: That should be synchronized set, as inputs can be closed by any threads. The `synchronized` on `close` does not cut it because it does not synchronize on this common utils object, but it synchronizes on the individual `IndexInput`. So two different index inputs closing at the same time would attempt to insert into this set concurrently.

Reply:
Right, good catch, fixed.